apache-maven/src/assembly/maven/conf/settings.xml

```diff
@@ -29,14 +29,14 @@ under the License.
  |
  |                 -s /path/to/user/settings.xml
  |
- |  2. Global Level. This settings.xml file provides configuration for all Maven
+ |  2. System Level. This settings.xml file provides configuration for all Maven
```

Comment:
I consider this a wrong as global. See https://cwiki.apache.org/confluence/display/MAVEN/Commandline+inheritance. It should be installation level.

Reply:
And what if I have defined on workstation `MAVEN_ARGS=-gs=/etc/maven/settings.xml`? Is that still "installation" level?

Reply:
I see that differently. We have "system" and "user" settings, and the utter difference between the two is precedence. Whatever from where they come, are they "default" or "override", precedence is all that matters.

Reply:
This is very much alike how we build "effective POM/models", Maven does not care from where the bits come (is current POM in local repo, or in reactor, is parent POM resolved from remote...), what mavan cares is HOW the final result assembled.

Reply:
Yes, correct. The default settings come with the Maven installation wich can differ from version to version. We don't have a such a notion "Maven system", but "Maven installation" we do.

Reply:
For me the fact that your "Maven installation" setting (that is the _default value_ for global settings) can be overridden by CLI switch `-gs` tells all. Plus, I like naming it more like "system" and not "global".

Again, "installation" is just and nothing more than a "sensible default". If you look at it as anything else, you are doing it wrong (or the -gs should not exist).

Reply:
FQN is "Maven System Level". We discussed here in context of Maven. So "Maven User Level" and "Maven System Level".

Reply:
One possibility would be to remove the explicit levels and just use list of files.
For example:
```
maven.install.conf = ${maven.conf}       # this property should not be defined 
maven.user.conf = ${user.home}/.m2
maven.project.conf = ${session.rootDirectory}/.mvn

# Toolchains
maven.install.toolchains = ${maven.install.conf}/toolchains.xml
maven.project.toolchains = ${maven.project.conf}/toolchains.xml
maven.user.toolchains = ${maven.user.conf}/toolchains.xml

maven.toolchains = ?${maven.install.toolchains}, \
                   ?${maven.project.toolchains}, \
                   ?${maven.user.toolchains}

# Extensions
maven.install.extensions = ${maven.install.conf}/extensions.xml
maven.project.extensions = ${maven.project.conf}/extensions.xml
maven.user.extensions = ${maven.user.conf}/extensions.xml

maven.extensions = ?${maven.install.extensions}, \
                   ?${maven.project.extensions}, \
                   ?${maven.user.extensions}
```

This would solve the problem and even allow the user to switch precedence very easily, as the list order would be the one used.

The `-gt` flag would set override the `maven.install.toolchains`.

Reply:
Even better, we can use the following:
```
maven.install.toolchains = ${cli:it:-${cli:gt:-${maven.install.conf}/toolchains.xml}}
maven.project.toolchains = ${maven.project.conf}/toolchains.xml
maven.user.toolchains    = ${cli:t:-${maven.user.conf}/toolchains.xml}
maven.toolchains = ?${maven.install.toolchains} \
                   ?${maven.project.toolchains} \
                   ?${maven.user.toolchains}
```

The `:-` is used to specify a default value, and `${cli:t}` will grab the value of the `-t` CLI flag.

Reply:
> FQN is "Maven System Level". We discussed here in context of Maven. So "Maven User Level" and "Maven System Level".

As in 'Maven System' and not Maven in the entire system (OS, etc)?

Reply:
@gnodet What I like is you are picking up the idea of https://github.com/apache/tomcat/blob/main/conf/catalina.properties#L33.

Honestly, I cannot decipher it: `${cli:it:-${cli:gt:-${maven.install.conf}/toolchains.xml}}` Takes too much brain muscles from first glance.

I wonder why those XML files cannot be optional by default and Maven would simply iterate over
```
?${maven.install.extensions} \
                   ?${maven.project.extensions} \
                   ?${maven.user.extensions}
```
At the end we define the semantics how a property is interpreted. Git or Subversion do search for a user and global config and it is not a failure if both isn't there. Same here. We could log the absence at debug level.

Reply:
I'm fine with removing the `?` and assuming all those files are optional.

But that does not address CLI parameters.  Right now, they are supported by dedicated code in MavenCli.  The code would check if `-t` is set, and set the user toolchain file accordingly, else compute the default.  It's one OR is the other, not both.  So we need a way to have default values in the syntax, that's what the following would do:
```
maven.user.toolchains.default = ${maven.user.conf}/toolchains.xml}
maven.user.toolchains         = ${cli:t:-${maven.user.toolchains.default}}
maven.toolchains = ${maven.install.toolchains} \
                   ${maven.project.toolchains} \
                   ${maven.user.toolchains}
```

Reply:
Actually, we may still need the `?`.  Currently, if you specify `-t foo.xml`, maven will fail if `foo.xml` file does not exist.  Whereas _default_ files can be safely ignored if they don't exist.  So we need to be able to know if a file was explicitely specified or just defaulted imho.  So maybe:
```
maven.user.toolchains         = ${cli:t:-?${maven.user.toolchains.default}}
```
Note that we could use `cli.t` instead of `cli:t`, this may be less confusing:
```
maven.user.toolchains         = ${cli.t:-?${maven.user.toolchains.default}}
```

Reply:
One problem is that project settings are validated differently than user or install settings...  If we have a single list, it will be hard to keep that information.  I think this is the only case where those xml files (toolchains, settings, extensions) are loaded differently.

Another problem which is not currently addressed, is the merge of `extensions.xml`, but that's a different topic...